src/Std/System/Uri.lean

```diff
@@ -0,0 +1,115 @@
+/-
+Copyright (c) 2019 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Lovett
+-/
+
+namespace System
+namespace Uri
+namespace UriEscape
+
+/- https://www.ietf.org/rfc/rfc3986.txt -/
+def Zero : UInt8 := '0'.toNat.toUInt8
+def Nine : UInt8 := '9'.toNat.toUInt8
+def Lettera : UInt8 := 'a'.toNat.toUInt8
+def Letterf : UInt8 := 'f'.toNat.toUInt8
+def LetterA : UInt8 := 'A'.toNat.toUInt8
+def LetterF : UInt8 := 'F'.toNat.toUInt8
+
+/-- Decode %HH escapings in the given string. Note that sometimes a consecutive
+sequence of multiple escapings can represet a utf-8 encoded sequence for
+a single unicode code point and these will also be decoded correctly. -/
+def decodeUri (uri : String) : String := Id.run do
+  let mut decoded : ByteArray := ByteArray.empty
+  let rawBytes := uri.toUTF8
+  let len := rawBytes.size
+  let mut i := 0
+  let percent := '%'.toNat.toUInt8
+  while i < len do
+    let c := rawBytes[i]!
+    let (result, pos) := if c == percent && i + 1 < len then
```

Comment:
Why are you using this pair instead of assigning to the mutable variables directly?

Reply:
I thought imperative code was bad :-) I found it looked a bit cleaner this way and it doesn't seem to change the performance is the compiler somehow optimizing away this pair object?

Reply:
Personally I think assigning directly to descriptively-named variables is clearer than using anonymous pairs. But it's fine either way.

Reply:
I see, then I think you might like the updated version that assigns directly to our existing variable `(decoded, i) := if c == percent && i + 1 < len then`  ...  as for anonymous pairs, I agree anonymous tuples can get out of hand but I think in this case it's pretty clear what's going on.  So do you mind if I leave it this way for now?  Were you expecting a performance hit from the pair, or does that being virtually free make sense to you?

Reply:
Ah, you said "But it's fine either way." so I'm resolving this conversation.